packages/financial-templates-lib/src/price-feed/UpunkPriceFeed.js

```diff
@@ -0,0 +1,168 @@
+    }
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  getLookback() {
+    return this.historicalLookback;
+  }
+
+  getPriceFeedDecimals() {
+    return this.priceFeedDecimals;
+  }
+
+  async update() {
+    // Approximate the first block from which we'll need price data from based on the
+    // lookback:
+    const lookbackWindow = this.historicalLookback;
+    const currentTime = await this.getTime();
+    const earliestLookbackTime = currentTime - lookbackWindow;
+    const latestBlockNumber = (await this.web3.eth.getBlock("latest")).number;
+    // Add cushion in case `averageBlockTimeSeconds` underestimates the seconds per block:
+    let lookbackBlocks = Math.ceil((this.bufferBlockPercent * lookbackWindow) / (await averageBlockTimeSeconds()));
+
+    let events = []; // Caches sorted events (to keep subsequent event queries as small as possible).
+    let fromBlock = Infinity; // Arbitrary initial value > 0.
+
+    // For loop continues until the start block hits 0 or the first event is before the earlest lookback time.
+    for (let i = 0; !(fromBlock === 0 || events[0]?.timestamp <= earliestLookbackTime); i++) {
+      // Uses latest unless the events array already has data. If so, it only queries _before_ existing events.
+      const toBlock = events[0] ? events[0].blockNumber - 1 : "latest";
+
+      // By taking larger powers of 2, this doubles the lookback each time.
+      fromBlock = Math.max(0, latestBlockNumber - lookbackBlocks * 2 ** i);
+
+      const newEvents = await this._getSortedPunkBoughtEvents(fromBlock, toBlock).then(newEvents => {
+        // Grabs the timestamps for all blocks, but avoids re-querying by .then-ing any cached blocks.
+        return Promise.all(
+          newEvents.map(event => {
+            // If there is nothing in the cache for this block number, add a new promise that will resolve to the block.
+            if (!this.blocks[event.blockNumber]) {
+              this.blocks[event.blockNumber] = this.web3.eth
+                .getBlock(event.blockNumber)
+                .then(block => ({ timestamp: block.timestamp, number: block.number }));
+            }
+
+            // Add a .then to the promise that sets the timestamp (and price) for this event after the promise resolves.
+            return this.blocks[event.blockNumber].then(block => {
+              event.timestamp = block.timestamp;
+              event.price = event.returnValues.value;
+              event.punkIndex = event.returnValues.punkIndex;
+              return event;
+            });
+          })
+        );
+      });
+
+      // Adds newly queried events to the array.
+      events = [...newEvents, ...events];
+    }
+
+    // If there are still no prices, return null to allow the user to handle the absence of data.
+    if (events.length === 0) {
+      this.events = [];
+      return;
+    }
+
+    // Filter out events where price is null.
+    this.events = events.filter(e => e.price !== null);
```

Comment:
Maybe you should filter out null prices before the `events.length === 0` check, so we don't try to call `this._computePrice` on an empty array